### PR TITLE
Implement initial PoC

### PR DESCRIPTION
**Pull request by mattkelly**
Here's an initial PoC implementation to kick things off. I have some more experiments locally, but this is boiled down enough for a high-level review. There are a few things that should definitely be updated before merging which are elaborated on below.

Tests should be run as follows:

1. Provision:

```
ginkgo -r -v -debug -failFast provision -- --template `pwd`/resources/templates/digital_ocean/ubuntu.json --cluster `pwd`/resources/clusters/digital_ocean/cluster.json --kubernetes-version 1.13.4
```

2. All other tests (pulls in Kubeconfig from last run):

```
ginkgo -r -v -debug -failFast tests
```

Some things to note:

1. The only state required between the provision and test runs currently is the Kubeconfig. Ideally, I would like to keep it this way. We can do this by attaching any state to the cluster itself - e.g. add a label for the template ID (so it can be cleaned up as part of a cleanup suite), or add a ConfigMap, or...
2. We should think about being smarter with the Kubeconfig - maybe take a `--kubeconfig` arg so the environment variable isn't required (which can make running multiple tests at once locally annoying), maybe skip provisioning if the Kubeconfig already exists, ...

The most important TODOs are:

1. Decide on a real resource template strategy and add those templates
2. Update Dockerfile to actually run tests
3. Add CI configuration file (discussed with Nick - I would strongly prefer to use CircleCI, but we'll run out of build hours quickly)
4. Fix the scale test to test for the proper things
5. Add some more useful provision tests (check Coordinator API, etc)

Other TODOS are:

- Lots of refactoring to deduplicate code etc
- Add upgrade tests
- Add node pool add/remove tests
- Other tests

Closes https://github.com/containership/cloud-test/issues/50.

### CODE REVIEW

File: Gopkg.toml [hunk 1]
```diff
@@ -0,0 +1,50 @@
+# Gopkg.toml example
```
Comment (mattkelly) [inline Gopkg.toml]: Delete comments

File: provision/provision_test.go [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+			Templates(context.OrganizationID).
+			Create(req)
+		Expect(err).NotTo(HaveOccurred())
+		Expect(resp).NotTo(BeNil())
+
+		// Set template ID in global context - should never be mutated after this
+		context.TemplateID = string(resp.ID)
+	})
+
+	It("should successfully initiate provisioning", func() {
```
Comment (nicktate) [inline provision/provision_test.go]: Is this not order dependent upon the completion if the previous test? (Right now we have the template_id) hard-coded, but I imagine we would want to:

1. Create template
2. Create cluster w/ template_id from 1

I could also be wrong in that these will run in a random order, does `ginkgo` only randomize `suites`?

File: provision/provision_test.go [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+
+// The provisionContext is different from the context required for other tests,
+// thus we split it out here
+type provisionContext struct {
+	ContainershipClientset cloud.Interface
+	KubernetesClientset    kubernetes.Interface
+
+	// AuthToken is only required because we can't pull the token back out of
+	// the Containership clientset to use it again
+	AuthToken string
```
[the remark this answers is not in the document]
Comment (mattkelly) [reply provision/provision_test.go]: I couldn't think of a use case outside of this fairly weird/hacky one, so it felt better to me to not add an accessor to the clientset.

File: provision/provision_test.go [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+}
+
+func TestProvision(t *testing.T) {
+	// Hook up gomega to ginkgo
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "Provision Suite")
+}
+
+var _ = SynchronizedBeforeSuite(func() []byte {
+	// Run only on first node
```
[the remark this answers is not in the document]
Comment (mattkelly) [reply provision/provision_test.go]: Very fair, this is a pretty bad comment :)

Ginkgo runs tests on "nodes" in a DAG basically. `SynchronizedBeforeSuite` will run the first function arg on the "entry" node of the graph and the second function arg on every other node after that. That's what the comments are poorly describing. Will update.

File: provision/provision_test.go [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+	context = &provisionContext{
+		ContainershipClientset: clientset,
+		AuthToken:              token,
+		KubeconfigFilename:     kubeconfigFilename,
+		OrganizationID:         constants.TestOrganizationID,
+	}
+
+	return nil
+}, func(_ []byte) {
+	// Run on all nodes after first one
```
[the remark this answers is not in the document]
Comment (mattkelly) [reply provision/provision_test.go]: See above

File: provision/provision_test.go (hunk 2 above)
Comment (mattkelly) [reply provision/provision_test.go]: Yeah ginkgo only randomizes suites (actually top-level `Describe` blocks, which was a point of a lot of confusion for me). You can tell it to randomize individual test cases if you want, but that'd be bad in our case for the reason you're alluding to. I touch on this a bit in the RFC.

File: resources/clusters/digital_ocean/cluster.json [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+         "implementation": "digitalocean",
+         "type": "csi"
+      },
+      {
+         "implementation": "digitalocean",
+         "type": "cloud_controller_manager"
+      }
+   ],
+   "provider_id": "08cd67a1-6837-487d-894d-d01827fbf840",
+   "template_id": "f0677fc9-3b79-4e14-9525-d65883a65057"
```
[the remark this answers is not in the document]
Comment (mattkelly) [reply resources/clusters/digital_ocean/cluster.json]: Yup, this should really be a go template (or similar). But there's no difference for now of just reading it in and replacing the value. Just super unclear if you look at the actual files.

I'd also strongly prefer that these are in yaml. Was just borrowing stuff from csctl (which I ultimately couldn't justify exporting on that side, at least not without thinking about it more).

**Comment by mattkelly** [comment]
1. Definitely.

2. It's important to note that currently the only requirement for running non-provision tests is that a proper kubeconfig file exists. It's not strictly necessary to run a provisioning test first.

3. I'm not sure I follow how this is better than dynamically pulling in test definitions (just cluster/template definitions, currently). Care to elaborate?

4. Yup, some suites will definitely be able to run in parallel - we just have to be careful about it. The way Kubernetes accomplishes this is by tagging tests that cannot run in parallel with `[Serial]`. They're then run in a separate invocation of ginkgo.

**Comment by mattkelly** [comment]
@nicktate and I discussed more via video call. Re: (3) I thought he was referring to putting all of provisioning in `SynchronizedBeforeSuite` as we had discussed in the RFC. This idea is different, and just saying that we might want to encode configuration variations in the test suite code itself rather than in e.g. even more command line flags. This could make a lot of sense. Going to think more on it.

File: tests/scale/scale_test.go [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+		var pool *types.NodePool
+		for _, p := range nodePools {
+			if *p.KubernetesMode == "worker" {
+				pool = &p
+				break
+			}
+		}
+		if pool == nil {
+			// There are no worker pools - that's fine
+			Skip("no worker pools to test scale up on")
```
Comment (mattkelly) [inline tests/scale/scale_test.go]: This isn't correct because it will only skip the current `It`, not the entire `Describe`.

**Comment by mattkelly** [comment]
Some further thoughts:

1. We can/should completely eliminate the cluster request files. It was easier to pull from a file for the PoC, but I think a much saner approach is to build out the cluster requests in code. Cluster requests are pretty simple:

- Name
- Environment
- Template ID
- Plugins

The only tricky thing is plugins. Currently the csctl (internal) APIs are only really friendly if coming from parsing command line flags. I think it should be easyish to export functions that make this easier for csctl consumers. I previously experimented with this but will have to revisit.

An alternative approach is to just shell out to csctl and forward some plugin flags all the way through, e.g. `--plugin-ccm=digitalocean@x.y.z`. The csctl command line usage is already super friendly here.

2. I thought more about the "each cluster configuration as a separate suite" thing and think that if we do (1), it would be cleaner and a lot less boilerplate to instead just have a lot of template files for all possible configurations. We can just flip on/off plugins using CLI flags to the ginkgo invocation.

**Comment by mattkelly** [comment]
CI is finally fixed 😅. This is good for a final review.

**Pull request merged**